File: day1.py
test_data = """1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet"""
with open("input_files/aoc23-1.txt") as file:
    data = file.read()

test_data2 = """two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen"""

def part1(data):
    results = []
    for word in data.split():
        digits = ""
        for letter in word:
            if letter.isdigit():
                digits += letter
        results.append(int(digits[0] + digits[-1]))

    return sum(results)


def word_to_digits(word: str) -> str:
    i = 0
    numbers = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
    digits = ""
    while i < len(word):
        if word[i].isdigit():
            digits += word[i]
        else:
            for n, number in enumerate(numbers):
                if word[i:].startswith(number):
                    digits += str(n+1)
        i += 1
    return digits

def part2(data):
    results = []
    for word in data.split():
        # print(word_to_digits(word))
        digits = word_to_digits(word)
        results.append(int(digits[0] + digits[-1]))
        # else:
            # results.append(int(digits))
    # print(results)
    # for result in results:
    #     if result > 99 or result < 0:
    #         print(f"warning {result}")
    return sum(results)




print(part2(data))
# print(replace_word_numbers("llnshf1"))
# print(word_to_digits("eighthree"))

File: day1.hs
import Data.List (isPrefixOf)

loadData :: String -> IO [String]
loadData f  = do
  contents <- readFile f
  let fileLines = lines contents
  return fileLines

testData = "1abc2\npqr3stu8vwx\na1b2c3d4e5f\ntreb7uchet\n"

extractDigits :: String -> String
extractDigits [] = []
extractDigits (x:xs)
    | "one" `isPrefixOf` (x:xs)   = '1' : extractDigits xs
    | "two" `isPrefixOf` (x:xs)   = '2' : extractDigits xs
    | "three" `isPrefixOf` (x:xs) = '3' : extractDigits xs
    | "four" `isPrefixOf` (x:xs)  = '4' : extractDigits xs
    | "five" `isPrefixOf` (x:xs)  = '5' : extractDigits xs
    | "six" `isPrefixOf` (x:xs)   = '6' : extractDigits xs
    | "seven" `isPrefixOf` (x:xs) = '7' : extractDigits xs
    | "eight" `isPrefixOf` (x:xs) = '8' : extractDigits xs
    | "nine" `isPrefixOf` (x:xs)  = '9' : extractDigits xs
    | x >= '1' && x <= '9'        = x : extractDigits xs
    | otherwise                   = extractDigits xs

getFstLstNumber :: String -> Int
getFstLstNumber xs = read $ head xs : [last xs]

numList :: [String] -> [Int]
numList = map getFstLstNumber

main :: IO ()
main = do
    inputdata <-loadData "input_files/aoc23-1.txt"
    print $ sum . numList $ map extractDigits inputdata

